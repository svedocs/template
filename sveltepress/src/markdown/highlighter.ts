import type { Highlighter } from '@sveltepress/vite'
import type { BundledTheme, CodeToHastOptions, HighlighterGeneric, ShikiTransformer } from 'shiki'
import type { BundledLanguage } from 'shiki/langs'
import type { DefaultThemeOptions } from 'virtual:sveltepress/theme-default'
import { env } from 'node:process'
import { createTransformerFactory } from '@shikijs/twoslash'
import { createTwoslasher, rendererFloatingSvelte } from '@sveltepress/twoslash'
import { LRUCache } from 'lru-cache'
import { createHighlighter } from 'shiki'
import { themeOptionsRef } from '../index.js'
import { processCommands } from './commands.js'

const DEFAULT_SUPPORT_LANGUAGES: any[] = ['svelte', 'sh', 'js', 'html', 'ts', 'md', 'css', 'scss']

const cache = new LRUCache<string, any>({ max: 200 })

const shikiHighlighterInstance: {
  value: null | HighlighterGeneric<BundledLanguage, BundledTheme>
  twoslashTransformer?: ShikiTransformer
} = {
  value: null,
}

const _highlighter: Highlighter = async (code, lang) => {
  const highlighterConfig = themeOptionsRef.value?.highlighter
  const darkTheme = highlighterConfig?.themeDark ?? 'night-owl'
  const lightTheme = highlighterConfig?.themeLight ?? 'vitesse-light'

  const shikiOptions: CodeToHastOptions<BundledLanguage, BundledTheme> = {
    lang,
    themes: {
      dark: darkTheme,
      light: lightTheme,
    },
    transformers: [],
  }
  if (shikiHighlighterInstance.twoslashTransformer) {
    shikiOptions.transformers?.push(
      shikiHighlighterInstance.twoslashTransformer,
    )
  }
  const { codeToHtml } = shikiHighlighterInstance.value!
  return codeToHtml(code, shikiOptions)
    .replace(/\{/g, '&#123;')
    .replace(/\}/g, '&#125;')
    .replace(/<!--svp-floating-snippet-start-->/g, '{#snippet floatingContent()}')
    .replace(/<!--svp-floating-snippet-end-->/g, '{/snippet}')
}

const highlighter: Highlighter = async (code, lang, meta) => {
  const cacheKey = JSON.stringify({ code, lang, meta })
  let cached
  if (env.NODE_ENV === 'development') {
    cached = cache.get(cacheKey)
    if (cached)
      return cached
  }
  const metaArray = (meta || '').split(' ')
  const containLineNumbers = metaArray.some(item => item.trim() === 'ln')
  const titleMeta = metaArray.find(item => item.startsWith('title='))
  const commandDoms: string[] = []
  const lines = code.split('\n')
  let noErrorsFirstLine: string | undefined
  if (lines[0] === '// @noErrors')
    noErrorsFirstLine = lines.shift()
  code = lines.map((line, i) => {
    const [commandDomsInOneLine, newLine] = processCommands(line, i, lines.length)
    commandDoms.push(...commandDomsInOneLine)
    return newLine
  }).join('\n')
  let title: string | undefined
  if (titleMeta)
    title = titleMeta.split('=')[1].replace(/(^")|("$)/g, '')
  if (noErrorsFirstLine)
    code = `${noErrorsFirstLine}\n${code}`

  cached = `
<div class="svp-code-block-wrapper">${title
  ? `<div class="svp-code-block--title">${title}</div>
`
  : ''}
  <div class="svp-code-block${containLineNumbers ? ' svp-code-block--with-line-numbers' : ''}">
    ${commandDoms.join('\n')}
    <!-- svelte-ignore a11y_no_noninteractive_tabindex -->
    ${await _highlighter(code, lang, meta)}
    <div class="svp-code-block--lang">
      ${lang}
    </div>
    <CopyCode />
    ${containLineNumbers ? `<div class="svp-code-block--line-numbers">${lines.map((_, i) => `<div class="svp-code-block--line-number-item">${i + 1}</div>`).join('\n')}</div>` : ''}
  </div>
</div>`
  if (env.NODE_ENV === 'development')
    cache.set(cacheKey, cached)
  return cached
}

export async function initHighlighter(highlighterConfig: DefaultThemeOptions['highlighter'] = {}) {
  const langs = highlighterConfig?.languages || DEFAULT_SUPPORT_LANGUAGES
  const darkTheme = highlighterConfig?.themeDark ?? 'night-owl'
  const lightTheme = highlighterConfig?.themeLight ?? 'vitesse-light'
  shikiHighlighterInstance.value = await createHighlighter({
    langs,
    themes: [darkTheme, lightTheme],
  })
  if (highlighterConfig.twoslash) {
    shikiHighlighterInstance.twoslashTransformer = createTransformerFactory(
      await createTwoslasher(highlighterConfig.twoslash === true
        ? {
            compilerOptions: {
              module: 199,
              moduleResolution: 99,
              jsx: 1,
              types: ['@sveltepress/vite/types', '@sveltepress/theme-default/types', '@sveltepress/theme-default/components', '@sveltejs/kit'],
            },
          }
        : highlighterConfig.twoslash),
    )(
      {
        langs: ['ts', 'tsx', 'svelte'],
        renderer: rendererFloatingSvelte(),
      },
    )
  }
}

export default highlighter
