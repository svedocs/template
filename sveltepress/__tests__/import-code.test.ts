import { readFileSync } from 'node:fs'
import { resolve } from 'node:path'
import { fileURLToPath } from 'node:url'
import { mdToSvelte } from '@sveltepress/vite'
import { describe, expect, it } from 'vitest'
import codeImport, { importRe } from '../src/markdown/code-import'
import highlighter, { initHighlighter } from '../src/markdown/highlighter'

const __dirname = fileURLToPath(new URL('.', import.meta.url))
const mdPath = resolve(__dirname, 'code-import.md')
const mdContent = readFileSync(mdPath, 'utf-8')

describe('code import', async () => {
  await initHighlighter()

  it('existing and non-existing file', async () => {
    expect(mdContent).toMatchInlineSnapshot(`
      "## Title

      @code(./fake-file.ts)

      @code(/src/index.ts,20,37)
      "
    `)

    const { code } = await mdToSvelte({
      mdContent,
      filename: mdPath,
      highlighter,
      remarkPlugins: [codeImport],
    })
    expect(code).toMatchSnapshot()
  })

  it('re tests', async () => {
    const matches = importRe.exec('@code(./foo/bar/Comp.svelte,10,20)')
    expect(matches).toMatchInlineSnapshot(`
      [
        "@code(./foo/bar/Comp.svelte,10,20)",
        "./foo/bar/Comp.svelte,10,20",
        undefined,
        undefined,
      ]
    `)
  })
})
