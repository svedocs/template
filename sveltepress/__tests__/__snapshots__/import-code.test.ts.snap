// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`code import > existing and non-existing file 2`] = `
"<h2>Title</h2>
<p>@code(./fake-file.ts)</p>

<div class="svp-code-block-wrapper">
  <div class="svp-code-block">
    
    <!-- svelte-ignore a11y_no_noninteractive_tabindex -->
    <pre class="shiki shiki-themes vitesse-light night-owl" style="background-color:#ffffff;--shiki-dark-bg:#011627;color:#393a34;--shiki-dark:#d6deeb" tabindex="0"><code><span class="line"><span style="color:#A0ADA0;--shiki-dark:#637777">//</span><span style="color:#A0ADA0;font-style:inherit;--shiki-dark:#637777;--shiki-dark-font-style:italic"> @noErrors</span></span>
<span class="line"><span style="color:#B07D48;--shiki-dark:#D6DEEB">value</span><span style="color:#AB5959;--shiki-dark:#C792EA">?:</span><span style="color:#B07D48;--shiki-dark:#D6DEEB"> DefaultThemeOptions</span></span>
<span class="line"><span style="color:#393A34;--shiki-dark:#D6DEEB">&#125; </span><span style="color:#999999;--shiki-dark:#C792EA">=</span><span style="color:#999999;--shiki-dark:#D6DEEB"> &#123;</span></span>
<span class="line"><span style="color:#998418;--shiki-dark:#D6DEEB">value</span><span style="color:#999999;--shiki-dark:#D6DEEB">: </span><span style="color:#AB5959;--shiki-dark:#82AAFF">undefined</span><span style="color:#999999;--shiki-dark:#D6DEEB">,</span></span>
<span class="line"><span style="color:#999999;--shiki-dark:#D6DEEB">&#125;</span></span>
<span class="line"></span>
<span class="line"><span style="color:#AB5959;--shiki-dark:#C792EA">const</span><span style="color:#59873A;font-style:inherit;--shiki-dark:#82AAFF;--shiki-dark-font-style:italic"> defaultTheme</span><span style="color:#999999;--shiki-dark:#7FDBCA">:</span><span style="color:#2E8F82;font-style:inherit;--shiki-dark:#FFCB8B;--shiki-dark-font-style:italic"> ThemeDefault</span><span style="color:#999999;--shiki-dark:#C792EA"> =</span><span style="color:#999999;--shiki-dark:#D9F5DD"> (</span><span style="color:#B07D48;font-style:inherit;--shiki-dark:#D7DBE0;--shiki-dark-font-style:italic">options</span><span style="color:#999999;--shiki-dark:#D9F5DD">)</span><span style="color:#999999;--shiki-dark:#C792EA"> =></span><span style="color:#999999;--shiki-dark:#C792EA"> &#123;</span></span>
<span class="line"><span style="color:#B07D48;font-style:inherit;--shiki-dark:#D6DEEB;--shiki-dark-font-style:italic">themeOptionsRef</span><span style="color:#999999;font-style:inherit;--shiki-dark:#C792EA;--shiki-dark-font-style:italic">.</span><span style="color:#B07D48;font-style:inherit;--shiki-dark:#BAEBE2;--shiki-dark-font-style:italic">value</span><span style="color:#999999;--shiki-dark:#C792EA"> =</span><span style="color:#B07D48;font-style:inherit;--shiki-dark:#D6DEEB;--shiki-dark-font-style:italic"> options</span></span>
<span class="line"></span>
<span class="line"><span style="color:#1E754F;font-style:inherit;--shiki-dark:#C792EA;--shiki-dark-font-style:italic">return</span><span style="color:#999999;--shiki-dark:#C792EA"> &#123;</span></span>
<span class="line"><span style="color:#998418;font-style:inherit;--shiki-dark:#C792EA;--shiki-dark-font-style:italic">  name</span><span style="color:#999999;--shiki-dark:#C792EA">:</span><span style="color:#B5695977;--shiki-dark:#D9F5DD"> '</span><span style="color:#B56959;--shiki-dark:#ECC48D">@sveltepress/theme-default</span><span style="color:#B5695977;--shiki-dark:#D9F5DD">'</span><span style="color:#999999;--shiki-dark:#C792EA">,</span></span>
<span class="line"><span style="color:#998418;font-style:inherit;--shiki-dark:#C792EA;--shiki-dark-font-style:italic">  globalLayout</span><span style="color:#999999;--shiki-dark:#C792EA">:</span><span style="color:#B5695977;--shiki-dark:#D9F5DD"> '</span><span style="color:#B56959;--shiki-dark:#ECC48D">@sveltepress/theme-default/GlobalLayout.svelte</span><span style="color:#B5695977;--shiki-dark:#D9F5DD">'</span><span style="color:#999999;--shiki-dark:#C792EA">,</span></span>
<span class="line"><span style="color:#998418;font-style:inherit;--shiki-dark:#C792EA;--shiki-dark-font-style:italic">  pageLayout</span><span style="color:#999999;--shiki-dark:#C792EA">:</span><span style="color:#B5695977;--shiki-dark:#D9F5DD"> '</span><span style="color:#B56959;--shiki-dark:#ECC48D">@sveltepress/theme-default/PageLayout.svelte</span><span style="color:#B5695977;--shiki-dark:#D9F5DD">'</span><span style="color:#999999;--shiki-dark:#C792EA">,</span></span>
<span class="line"><span style="color:#59873A;font-style:inherit;--shiki-dark:#82AAFF;--shiki-dark-font-style:italic">  vitePlugins</span><span style="color:#999999;--shiki-dark:#C792EA">:</span><span style="color:#AB5959;font-style:inherit;--shiki-dark:#C792EA;--shiki-dark-font-style:italic"> async</span><span style="color:#999999;--shiki-dark:#D9F5DD"> (</span><span style="color:#B07D48;font-style:inherit;--shiki-dark:#D7DBE0;--shiki-dark-font-style:italic">corePlugin</span><span style="color:#999999;--shiki-dark:#D9F5DD">)</span><span style="color:#999999;--shiki-dark:#C792EA"> =></span><span style="color:#999999;--shiki-dark:#C792EA"> &#123;</span></span>
<span class="line"><span style="color:#AB5959;--shiki-dark:#C792EA">    const</span><span style="color:#B07D48;font-style:inherit;--shiki-dark:#82AAFF;--shiki-dark-font-style:italic"> plugins</span><span style="color:#999999;--shiki-dark:#C792EA"> =</span><span style="color:#999999;--shiki-dark:#D6DEEB"> [</span></span>
<span class="line"><span style="color:#999999;--shiki-dark:#7FDBCA">      ...</span><span style="color:#1E754F;font-style:inherit;--shiki-dark:#C792EA;--shiki-dark-font-style:italic">await</span><span style="color:#59873A;font-style:inherit;--shiki-dark:#82AAFF;--shiki-dark-font-style:italic"> createPreCorePlugins</span><span style="color:#999999;--shiki-dark:#D6DEEB">(</span><span style="color:#B07D48;font-style:inherit;--shiki-dark:#D6DEEB;--shiki-dark-font-style:italic">options</span><span style="color:#999999;--shiki-dark:#D6DEEB">),</span></span>
<span class="line"><span style="color:#B07D48;font-style:inherit;--shiki-dark:#D6DEEB;--shiki-dark-font-style:italic">      corePlugin</span><span style="color:#999999;--shiki-dark:#D6DEEB">,</span></span>
<span class="line"><span style="color:#999999;--shiki-dark:#D6DEEB">    ]</span></span>
<span class="line"><span style="color:#1E754F;--shiki-dark:#C792EA">    if</span><span style="color:#999999;--shiki-dark:#D6DEEB"> (</span><span style="color:#B07D48;font-style:inherit;--shiki-dark:#D6DEEB;--shiki-dark-font-style:italic">options</span><span style="color:#999999;font-style:inherit;--shiki-dark:#C792EA;--shiki-dark-font-style:italic">?.</span><span style="color:#B07D48;font-style:inherit;--shiki-dark:#BAEBE2;--shiki-dark-font-style:italic">pwa</span><span style="color:#999999;--shiki-dark:#D6DEEB">)</span><span style="color:#999999;--shiki-dark:#C792EA"> &#123;</span></span></code></pre>
    <div class="svp-code-block--lang">
      ts
    </div>
    <CopyCode />
    
  </div>
</div>"
`;
